benches/benchmarks.rs

```diff
@@ -172,12 +178,15 @@ fn big_16(c: &mut Criterion) {
     });
     group.bench_function("memchr", |b| {
         b.iter(|| {
-            memchr::memchr3(b'A', b'B', b'C', haystack.as_bytes())
-                .or_else(|| memchr::memchr3(b'D', b'E', b'F', haystack.as_bytes()))
-                .or_else(|| memchr::memchr3(b'G', b'H', b'I', haystack.as_bytes()))
-                .or_else(|| memchr::memchr3(b'J', b'K', b'L', haystack.as_bytes()))
-                .or_else(|| memchr::memchr3(b'M', b'N', b'O', haystack.as_bytes()))
-                .or_else(|| memchr::memchr(b'P', haystack.as_bytes()))
+            let indexes = [
+                memchr::memchr3(b'A', b'B', b'C', haystack.as_bytes()),
+                memchr::memchr3(b'D', b'E', b'F', haystack.as_bytes()),
+                memchr::memchr3(b'G', b'H', b'I', haystack.as_bytes()),
+                memchr::memchr3(b'J', b'K', b'L', haystack.as_bytes()),
+                memchr::memchr3(b'M', b'N', b'O', haystack.as_bytes()),
+                memchr::memchr(b'P', haystack.as_bytes()),
+            ];
+            indexes.iter().copied().min_by_key(|x| x.unwrap_or(usize::MAX)).unwrap()
```

Comment:
I think this can be `a.into_iter().flatten().min()`. I'd recommend using the same array pattern in the `memchr` usage above, instead of the `match`.

This commit, my suggestion, and the original commit that introduced this benchmark can all be squashed — no reason to have the intermediate state committed forever.

Reply:
I wonder if it would make it cleaner to convert to bytes at the beginning:

```rust
let haystack = haystack.as_bytes();
let indexes = [ /* ... */ ];
```

Reply:
Amusingly, I'm wondering if you could make a macro that accepts an arbitrary number of arguments and builds this collection of `memchr{3,2,1}` searches for you. If memchr truly is that much faster, that may end up being a wholesale replacement for this crate 😬

Reply:
Will need to go to the 2021 edition to use `array.into_iter()` (or use `array.iter().copied()` since it's Option<usize>), is that okay?

Reply:
Ah, one reason it wouldn't be a a wholesale replacement: if you run two memchrs, you can end up in the case where one finds a result immediately, and the other has to scan the whole string to find nothing.

Like `[memchr3(b'a', 'b', 'c', haystack), memchr(b'!', haystack)]` on a string of 1 GiB of `a`s, will give `[Some(1), None]` only after searching through the whole string. The current benchmarks give the best possible case for multiple memchrs, where all memchrs will scan the whole string and not find anything, and one will find the last character.

Multiple memchrs are only really a replacement if you want to traverse the whole string and find all instances of the specified chars. You also need to use the memchr iterators, and merge them, to avoid duplicating work rather than starting each search again after the first item is found (like in tafia/quick-xml#664)

Reply:
> to use `array.into_iter()`

`IntoIterator::into_iter([...]);` works in all editions, but I'm not really worried about that maximum performance. It could be `iter().copied().flatten()` for succinctness.

Reply:
> The current benchmarks give the best possible case for multiple memchrs

That's the *worst* case for memchr, right? Which would be the best to make jetscii look good, right?

Otherwise, you could at least restrict subsequent searches to not look past any previously found indexes.

Reply:
No, I believe it's the best case for memchr: any method has to scan the whole haystack anyway, so it'll never be looking past the first character in the set we're interested in.

> you could at least restrict subsequent searches to not look past any previously found indexes.

True, but it's always possible every memchr search but the last will return None, and have to scan the whole string, and the last memchr search could find an item at index 0.